Cache regexps generated from acronym_regex

NickLaMuro opened the pull request:
Summary
-------

The following line from `String#camelize`:

```ruby
string = string.sub(/^(?:#{inflections.acronym_regex}(?=\b|[A-Z_])|\w)/) { |match| match.downcase }
```

and the following line from `String#camelize`:

```ruby
word.gsub!(/(?:(?<=([A-Za-z\d]))|\b)(#{inflections.acronym_regex})(?=\b|[^a-z])/) { "#{$1 && '_'.freeze }#{$2.downcase}" }#{$2.downcase}" }
```

Both generate the same regexep in the first part of the `.sub`/`.gsub` method calls every time the function is called, creating an extra object allocation each time.  The value of `acronym_regex` only changes if the user decides add an acronym to the current set of inflections and apends another string on the the regexp generated here, but beyond that it remains relatively static.

This has been around since acronym support was introduced back in 2011 in https://github.com/rails/rails/pull/1648.

To avoid re-generating these strings every time these methods are called, cache the values of these regular expressions in the `ActiveSupport::Inflector::Inflections` instance, making it so these regular expressions are only generated once, or when the acronym's are added to.


Other Information
-----------------

Benchmarks against this code, and the most recently released version of `activesupport` can be run using this gist:

https://gist.github.com/NickLaMuro/a91406e512a59c5b676f110ce968680b

When running it using the following for the current version of `ActiveSupport`, and the changes after this patch, there were 20% less object allocations, and the script ran 20% - 30% faster.  Sample output can be found below:


**Before:**

```console
$ ruby test_acronym_fixes.rb --install-gems --silent-bundler
Total allocated: 59551503 bytes (510009 objects)
Total retained:  320 bytes (2 objects)

allocated memory by gem
-----------------------------------
  44970885  activesupport-5.1.4
  14580618  other

allocated memory by file
-----------------------------------
  41770716  /Users/nicklamuro/.gem/ruby/2.3.5/gems/activesupport-5.1.4/lib/active_support/inflector/methods.rb
  14580618  test_acronym_fixes.rb
   3200169  /Users/nicklamuro/.gem/ruby/2.3.5/gems/activesupport-5.1.4/lib/active_support/core_ext/regexp.rb

allocated memory by location
-----------------------------------
  19500040  /Users/nicklamuro/.gem/ruby/2.3.5/gems/activesupport-5.1.4/lib/active_support/inflector/methods.rb:93
   9780169  /Users/nicklamuro/.gem/ruby/2.3.5/gems/activesupport-5.1.4/lib/active_support/inflector/methods.rb:95
   8090169  test_acronym_fixes.rb:84
   7690169  /Users/nicklamuro/.gem/ruby/2.3.5/gems/activesupport-5.1.4/lib/active_support/inflector/methods.rb:74


Timings
-------
real    0m5.346s
user    0m5.210s
sys     0m0.130s
cuser   0m0.000s
csys    0m0.000s

Memory used MB: 429.05859375
```


**After:**

```console
$ ruby test_acronym_fixes.rb --install-gems --silent-bundler --local
Total allocated: 40051632 bytes (420009 objects)
Total retained:  320 bytes (2 objects)

allocated memory by gem
-----------------------------------
  25471014  activesupport/lib
  14580618  other

allocated memory by file
-----------------------------------
  22270845  /Users/nicklamuro/code/public_forks/rails-dev-box/rails/activesupport/lib/active_support/inflector/methods.rb
  14580618  test_acronym_fixes.rb
   3200169  /Users/nicklamuro/code/public_forks/rails-dev-box/rails/activesupport/lib/active_support/core_ext/regexp.rb

allocated memory by location
-----------------------------------
   9780169  /Users/nicklamuro/code/public_forks/rails-dev-box/rails/activesupport/lib/active_support/inflector/methods.rb:99
   8090169  test_acronym_fixes.rb:84
   7690169  /Users/nicklamuro/code/public_forks/rails-dev-box/rails/activesupport/lib/active_support/inflector/methods.rb:77
   6490449  test_acronym_fixes.rb:83


Timings
-------
real    0m4.309s
user    0m4.210s
sys     0m0.090s
cuser   0m0.000s
csys    0m0.000s

Memory used MB: 364.76953125
```

NickLaMuro committed:
Cache regexps generated from acronym_regex

The Problem
-----------

The following line from `String#camelize`:

  string = string.sub(/^(?:#{inflections.acronym_regex}(?=\b|[A-Z_])|\w)/) { |match| match.downcase }

and the following line from `String#camelize`:

  word.gsub!(/(?:(?<=([A-Za-z\d]))|\b)(#{inflections.acronym_regex})(?=\b|[^a-z])/) { "#{$1 && '_'.freeze }#{$2.downcase}" }#{$2.downcase}" }

Both generate the same regexep in the first part of the `.sub`/`.gsub`
method calls every time the function is called, creating an extra object
allocation each time.  The value of `acronym_regex` only changes if the
user decides add an acronym to the current set of inflections and apends
another string on the the regexp generated here, but beyond that it
remains relatively static.

This has been around since acronym support was introduced back in 2011
in PR#1648.

Proposed Solution
-----------------
To avoid re-generating these strings every time these methods are
called, cache the values of these regular expressions in the
`ActiveSupport::Inflector::Inflections` instance, making it so these
regular expressions are only generated once, or when the acronym's are
added to.

Other notable changes is the attr_readers are nodoc'd, as they shouldn't
really be public APIs for users.  Also, the new method,
define_acronym_regex_patterns, is the only method in charge of
manipulating @acronym_regex, and initialize_dup also makes use of that
new change.

** Note about fix for non-deterministic actionpack test **

With the introduction of `@acronym_underscore_regex` and
`@acronym_camelize_regex`, tests that manipulated these for a short
time, then reset them could caused test failures to happen.  This
happened because the previous way we reset the `@acronyms` and
`@acronym_regex` was the set them using #instance_variable_set, which
wouldn't run the #define_acronym_regex_patterns method.

This has now been introduced into the actionpack tests to avoid this
failure.

NickLaMuro commented:
@tenderlove a mutual co-worker/previous co-worker thought you might find this patch interesting.

matthewd commented:
Hey @NickLaMuro!

This does sound like a good idea. We're already caching a semi-compiled regex, so it makes a lot of sense to just cache the two places it gets used instead.

The new accessors should be nodoc... and let's deprecate the `acronym_regex` reader too -- that seems to have no reason to stay, but as it's [accidentally?] documented, we can't just drop it silently.

Then let's make `define_acronym_regex_patterns` exclusively responsible for `@acronym_regex`, by teaching it the null pattern from `initialize`.

And finally, `initialize_dup` looks deserving of some attention: it shouldn't touch `@acronym_regex` directly, but probably does want to call `define_acronym_regex_patterns`.

NickLaMuro commented:
@matthewd I agree with all of your points, and thanks for reviewing.  I will look into doing those items in the next day or so when time permits.  Thanks again!

NickLaMuro committed:
Deprecate ActiveSupport::Inflector#acronym_regex

To be removed in Rails 6.0 (default for the deprecate helper).  Code
moved around as well for the ActiveSupport::Deprecation modules, since
it was dependent on ActiveSupport::Inflector being loaded for it to
work.  By "lazy loading" the Inflector code from within the Deprecation
code, we can require ActiveSupport::Deprecation from
ActiveSupport::Inflector and not get a circular dependency issue.

NickLaMuro commented:
@matthewd Hey, sorry for the delay on this, but I think I opened this going into a week of PTO, and it just fell off my priority list with the weird week I had last week.

Pretty sure I have addressed [your concerns](https://github.com/rails/rails/pull/30782#issuecomment-334527892), but I was doing some `git blame`ing to figure out how to do some of it, so hopefully how I did it was acceptable.  Let me know if there is anything you would like me to change.

rafaelfranca commented:
r? @matthewd

rafaelfranca reviewed (approved):
The code looks good to me but the commits could be squashed in two different commits (or three):

One of the caching addition, one for the deprecation of acronym_regex, one of fixing the non-deterministic test.

NickLaMuro commented:
> The code looks good to me but the commits could be squashed in two different commits (or three)

@rafaelfranca That is my general goto, but avoided it since your bot suggested that I didn't for the reviewer's sake 😄 

But I have no problem squashing commits, so will do.  That said, looks like I have some tests to fix as well, so I will also address that as well.

NickLaMuro commented:
Looks like there is a bit of a race condition here.  For me to use `ActiveSupport::Deprecation` for `ActiveSupport::Inflector`, `ActiveSupport::Inflector` needs to be already initialized (see [`constant_accessor.rb`](https://github.com/rails/rails/blob/589dd0f/activesupport/lib/active_support/deprecation/constant_accessor.rb) and [`proxy_wrappers.rb`](https://github.com/rails/rails/blob/589dd0f/activesupport/lib/active_support/deprecation/proxy_wrappers.rb)).  Going to try to figure out a way around this, but since this might already be a solved problem, letting those following along at home know about the issue incase they have input.

NickLaMuro commented:
Looks my solution to the above, while it works, adds a really nasty warning:

```
/home/travis/build/rails/rails/activesupport/lib/active_support/deprecation/constant_accessor.rb:3: warning: loading in progress, circular require considered harmful - /home/travis/build/rails/rails/activesupport/lib/active_support/inflector/methods.rb
	from /home/travis/build/rails/rails/vendor/bundle/ruby/2.4.0/gems/rake-12.1.0/lib/rake/rake_test_loader.rb:5:in `<main>'
	from /home/travis/build/rails/rails/vendor/bundle/ruby/2.4.0/gems/rake-12.1.0/lib/rake/rake_test_loader.rb:5:in `select'
	from /home/travis/build/rails/rails/vendor/bundle/ruby/2.4.0/gems/rake-12.1.0/lib/rake/rake_test_loader.rb:17:in `block in <main>'
	from /home/travis/build/rails/rails/vendor/bundle/ruby/2.4.0/gems/rake-12.1.0/lib/rake/rake_test_loader.rb:17:in `require'
...
```

Going to look at a way to make that not a thing.  Might not be able to use `ActiveSupport::Deprecation` for this, but we will see.

NickLaMuro commented:
Just pushed a rebase of the deprecation, but it is pretty gross... and it is definitely something I am proud of.  Hopefully it works, but I am open to more graceful suggestions (previous attempt that got rebased out is also available for viewing above).

matthewd commented:
Can we move those requires into `DeprecatedConstantAccessor.included` and `DeprecatedConstantProxy#initialize`, respectively?

At a glance, it looks like that should mean they'll both still happen at deprecation-definition time (as opposed to when the deprecation is encountered at "runtime"), but only when someone tries to deprecate a constant. That'd mean we still can't use it to deprecate a *constant* in the inflector, but that seems a much narrower limitation to live with.

NickLaMuro commented:
@matthewd Sorry for the delay in response.  Your 4am suggestion last time made it so it fell off my radar this week 😄 

Anyway, looks like your suggestion should work (have run some of the troublesome tests locally and they worked just fine).  Thanks for the excellent suggestion, I have implemented it, and hopefully the tests still pass.

matthewd commented on actionpack/test/dispatch/request_test.rb:
```diff
@@ -763,7 +763,9 @@ class RequestMethod < BaseRequestTest
-    existing_acronym_regex = ActiveSupport::Inflector.inflections.acronym_regex.dup
+    existing_acronym_regex = assert_deprecated do
+      ActiveSupport::Inflector.inflections.acronym_regex.dup
+    end
```
Don't need this variable anymore

matthewd commented on activesupport/lib/active_support/inflector/inflections.rb:
```diff
@@ -66,17 +67,23 @@ def self.instance(locale = :en)
 
-      attr_reader :plurals, :singulars, :uncountables, :humans, :acronyms, :acronym_regex
+      attr_reader :plurals, :singulars, :uncountables, :humans,
+                  :acronyms, :acronym_regex
```
Keep this on one line as we're not modifying it

matthewd commented on activesupport/test/inflector_test.rb:
```diff
@@ -224,6 +224,12 @@ def test_underscore_acronym_sequence
     assert_equal("json_html_api", ActiveSupport::Inflector.underscore("JSONHTMLAPI"))
   end
 
+  def test_acronum_regexp_is_deprecated
```
*acronym

matthewd reviewed (approved):
Looks great!

NickLaMuro commented:
Oops, I didn't see that this PR had failing tests until just now.

That said, it might have been a error on travis since I just see "An error occurred while generating the build script." for the relevant failing build.  That said, I will try and take a look locally before kicking Travis and having retry all the jobs, hopefully before the end of the day.

NickLaMuro commented:
Sorry for the delay on this (again...), but looks like this might just be a travis hiccup, kicking the build (closing and re-opening).

NickLaMuro commented:
🎉 😁

matthewd commented:
🎉

Sorry, I hadn't seen your push since my previous comment. 😞

NickLaMuro commented:
No worries, I wasn't sure that the test failures that happened were due to my changes or a travis hiccup, so I wanted to verify that prior to you merging anyway, and only got around to that today.

Thanks again!
